include/pangolin/video/drivers/thread.h

```diff
@@ -0,0 +1,234 @@
+
+    FixSizeBuffersQueue() {}
+
+    ~FixSizeBuffersQueue() {
+        // Deallocate everything.
+        std::lock_guard<std::mutex> vlock(vMtx);
+        while(validBuffers.size() > 0){
+            delete[] validBuffers.front();
+            validBuffers.pop_front();
+        }
+        std::lock_guard<std::mutex> elock(eMtx);
+        while(emptyBuffers.size() > 0){
+            delete[] emptyBuffers.front();
+            emptyBuffers.pop_front();
+        }
+    }
+
+    void init(unsigned int num, unsigned int sizeBytes) {
+        maxNumBuffers = num;
+        bufferSizeBytes = sizeBytes;
+        // lock queue
+        std::lock_guard<std::mutex> vlock(vMtx);
+        std::lock_guard<std::mutex> elock(eMtx);
+
+        // Put back any valid buffer to the available buffers queue.
+        while(validBuffers.size() > 0){
+            emptyBuffers.push_back(validBuffers.front());
+            validBuffers.pop_front();
+        }
+        // Allocate buffers
+        while(emptyBuffers.size() < maxNumBuffers) {
+            emptyBuffers.push_back(new unsigned char[bufferSizeBytes]);
+        }
+    }
+
+    BufPType getNewest() {
+        std::lock_guard<std::mutex> vlock(vMtx);
+        std::lock_guard<std::mutex> elock(eMtx);
+        if(validBuffers.size() == 0) {
+            // Empty queue.
+            return 0;
+        } else {
+            // Requeue all but newest buffers.
+            while(validBuffers.size() > 1) {
+                emptyBuffers.push_back(validBuffers.front());
+                validBuffers.pop_front();
+            }
+            // Return newest buffer.
+            BufPType bp = validBuffers.front();
+            validBuffers.pop_front();
+            return bp;
+        }
+    }
+
+    BufPType getNext() {
+        std::lock_guard<std::mutex> vlock(vMtx);
+        if(validBuffers.size() == 0) {
+            // Empty queue.
+            return 0;
+        } else {
+            // Return oldest buffer.
+            BufPType bp = validBuffers.front();
+            validBuffers.pop_front();
+            return bp;
+        }
+    }
+
+    BufPType getFreeBuffer() {
+        std::lock_guard<std::mutex> vlock(vMtx);
+        std::lock_guard<std::mutex> elock(eMtx);
+        if(emptyBuffers.size() > 0) {
+            // Simply get a free buffer from the free buffers list.
+            BufPType bp = emptyBuffers.front();
+            emptyBuffers.pop_front();
+            return bp;
+        } else {
+            if(validBuffers.size() == 0) {
+                // Queue not yet initialized.
+                return 0;
+            } else {
+                // No free buffers return oldest among the valid buffers.
+                BufPType bp = validBuffers.front();
+                validBuffers.pop_front();
+                return bp;
+            }
+        }
+    }
+
+    void addValidBuffer(BufPType bp) {
+        // Add buffer to valid buffers queue.
+        std::lock_guard<std::mutex> vlock(vMtx);
```

Comment:
lock_guard and unique_lock are also part of boost so I simply added a compat/locks.h header to take care of the c++11 or boost inclusion following other examples in boostd.